noisy-simulator/src/density_matrix_simulator.rs

```diff
@@ -0,0 +1,346 @@
+    }
+
+    /// Returns a reference to the vector containing the density matrix's data.
+    pub fn data(&self) -> &ComplexVector {
+        &self.data
+    }
+
+    /// Returns dimension of the matrix. E.g.: If the matrix is 5 x 5, then dimension is 5.
+    pub fn dimension(&self) -> usize {
+        self.dimension
+    }
+
+    /// Returns the number of qubits in the system.
+    pub fn number_of_qubits(&self) -> usize {
+        self.number_of_qubits
+    }
+
+    /// Returns `true` if the matrix is Hermitian.
+    fn is_hermitian(&self) -> bool {
+        for row in 0..self.dimension {
+            for col in 0..self.dimension {
+                let elt = self.data[self.dimension * row + col];
+                let mirror_elt = self.data[self.dimension * col + row];
+                if (elt.re - mirror_elt.re).abs() > TOLERANCE
+                    || (elt.im + mirror_elt.im).abs() > TOLERANCE
+                {
+                    return false;
+                }
+            }
+        }
+        true
+    }
+
+    /// Returns `true` if the trace of the matrix is 1.
+    fn is_normalized(&self) -> bool {
+        (self.trace() - 1.0).abs() <= TOLERANCE
+    }
+
+    /// Returns the trace of the matrix. The trace is the sum of the diagonal entries of a matrix.
+    fn trace(&self) -> f64 {
+        let mut trace: Complex<f64> = Complex::ZERO;
+        for idx in 0..self.dimension {
+            trace += self.data[(self.dimension + 1) * idx];
+        }
+        assert!(
+            trace.im <= TOLERANCE,
+            "state trace is not real, imaginary part is {}",
+            trace.im
+        );
+        trace.re
+    }
+
+    /// Return theoretical change in trace due to operations that have been applied so far.
+    /// In reality, the density matrix is always renormalized after instruments / operations
+    /// have been applied.
+    pub fn trace_change(&self) -> f64 {
+        self.trace_change
+    }
+
+    /// Renormalizes the matrix such that the trace is 1.
+    fn renormalize(&mut self) -> Result<(), Error> {
+        self.renormalize_with_trace(self.trace())
+    }
+
+    /// Renormalizes the matrix such that the trace is 1. Uses a precomputed `trace`.
+    fn renormalize_with_trace(&mut self, trace: f64) -> Result<(), Error> {
+        if trace < TOLERANCE {
+            return Err(Error::ProbabilityZeroEvent);
+        }
+        self.trace_change *= trace;
+        let renormalization_factor = 1.0 / trace;
+        for entry in self.data.iter_mut() {
+            *entry *= renormalization_factor;
+        }
+        Ok(())
+    }
+
+    /// Applies the operation matrix to the target qubits.
+    fn apply_operation_matrix(
+        &mut self,
+        operation_matrix: &SquareMatrix,
+        qubits: &[usize],
+    ) -> Result<(), Error> {
+        // TODO [Research]: Figure out why they do this qubits_expanded thing.
+        let mut qubits_expanded = Vec::with_capacity(2 * qubits.len());
+        for id in qubits {
```

Comment:
What was the result of the research on this TODO?

Reply:
Still haven't figured out